activesupport/lib/active_support/message_pack/extensions.rb

```diff
@@ -0,0 +1,277 @@
+          unpacker: method(:read_complex),
+          recursive: true
+
+        factory.register_type 5, Range,
+          packer: method(:write_range),
+          unpacker: method(:read_range),
+          recursive: true
+
+        factory.register_type 6, ActiveSupport::HashWithIndifferentAccess,
+          packer: method(:write_hash_with_indifferent_access),
+          unpacker: method(:read_hash_with_indifferent_access),
+          recursive: true
+
+        factory.register_type 7, Set,
+          packer: method(:write_set),
+          unpacker: method(:read_set),
+          recursive: true
+
+        factory.register_type 8, Time,
+          packer: method(:write_time),
+          unpacker: method(:read_time),
+          recursive: true
+
+        factory.register_type 9, DateTime,
+          packer: method(:write_datetime),
+          unpacker: method(:read_datetime),
+          recursive: true
+
+        factory.register_type 10, Date,
+          packer: method(:write_date),
+          unpacker: method(:read_date),
+          recursive: true
+
+        factory.register_type 11, ActiveSupport::TimeWithZone,
+          packer: method(:write_time_with_zone),
+          unpacker: method(:read_time_with_zone),
+          recursive: true
+
+        factory.register_type 12, ActiveSupport::TimeZone,
+          packer: method(:dump_time_zone),
+          unpacker: method(:load_time_zone)
+
+        factory.register_type 13, ActiveSupport::Duration,
+          packer: method(:write_duration),
+          unpacker: method(:read_duration),
+          recursive: true
+
+        factory.register_type 14, URI::Generic,
+          packer: :to_s,
+          unpacker: URI.method(:parse)
+
+        factory.register_type 15, IPAddr,
+          packer: :to_s,
+          unpacker: :new
+
+        factory.register_type 16, Pathname,
+          packer: :to_s,
+          unpacker: :new
+
+        factory.register_type 17, Regexp,
+          packer: :to_s,
+          unpacker: :new
+
+        factory.register_type 18, Module,
+          packer: method(:dump_module),
+          unpacker: method(:load_module)
+
+        factory.register_type 127, Object,
+          packer: method(:write_object),
+          unpacker: method(:read_object),
+          recursive: true
+
+        factory
+      end
+
+      LOAD_WITH_MSGPACK_EXT = 0
+      LOAD_WITH_JSON_CREATE = 1
+
+      def write_object(object, packer)
+        if object.class.respond_to?(:from_msgpack_ext)
+          packer.write(LOAD_WITH_MSGPACK_EXT)
+          write_module(object.class, packer)
+          packer.write(object.to_msgpack_ext)
+        elsif object.class.respond_to?(:json_create)
+          packer.write(LOAD_WITH_JSON_CREATE)
+          write_module(object.class, packer)
+          packer.write(object.as_json)
+        elsif object.respond_to?(:serializable_hash)
+          packer.write(object.serializable_hash)
+        elsif object.respond_to?(:as_json)
+          packer.write(object.as_json)
```

Comment:
For a cache serializer that makes sense, I agree.  But as a replacement for `JSON`, I think this would be a core use case.

Reply:
Yeah, but I also think `JSON` shouldn't do that :p https://github.com/flori/json/pull/519

I strongly believe that implicit type coercion in serializers is a huge source of bugs.

Reply:
I hadn't thought of it like that.  It does make a certain amount of sense.  And, if needed, it would be much easier to add the behavior later than it would be to take it away.